crates/sargon/src/system/sargon_os/pre_authorization/sargon_os_pre_authorization_status.rs

```diff
@@ -0,0 +1,360 @@
+use crate::prelude::*;
+use std::time::Duration;
+
+// ==================
+// Poll PreAuthorization Status (Public)
+// ==================
+impl SargonOS {
+    /// Polls the status of a `SubintentHash` until it is either `Success` or `Expired`.
+    pub async fn poll_pre_authorization_status(
+        &self,
+        intent_hash: SubintentHash,
+        expiration: DappToWalletInteractionSubintentExpiration,
+    ) -> Result<PreAuthorizationStatus> {
+        let (status, _) = self
+            .poll_pre_authorization_status_with_delays(intent_hash, expiration)
+            .await?;
+
+        Ok(status)
+    }
+}
+
+// ==================
+// Poll PreAuthorization Status (Internal)
+// ==================
+impl SargonOS {
+    /// Polls the state of a Subintent until it is submitted
+    ///
+    /// It returns the `PreAuthorizationStatus`, but also the list of delays between each poll.
+    async fn poll_pre_authorization_status_with_delays(
+        &self,
+        intent_hash: SubintentHash,
+        expiration: DappToWalletInteractionSubintentExpiration,
+    ) -> Result<(PreAuthorizationStatus, Vec<u64>)> {
+        let network_id = self.current_network_id()?;
+        let gateway_client = GatewayClient::new(
+            self.clients.http_client.driver.clone(),
+            network_id,
+        );
+
+        // We are going to play safe and leave an extra second to make sure we check the status one second after it has theoretically expired.
+        // This is to avoid considering expired a subintent that got committed in the last instant.
+        let seconds_until_expiration =
+            self.seconds_until_expiration(expiration) + 1;
+        let mut delays: Vec<u64> = vec![];
+
+        let mut delay_duration = POLLING_DELAY_INCREMENT;
+
+        loop {
+            // Check the subintent status
+            let response = gateway_client
+                .subintent_status(SubintentStatusRequest::new(
+                    intent_hash.to_string(),
+                ))
+                .await?;
+
+            match response.subintent_status {
+                SubintentStatus::CommittedSuccess => {
+                    // If it has been committed, we consider it a `Success`.
+                    let transaction_intent_hash = response
+                        .finalized_at_transaction_intent_hash
+                        .ok_or(CommonError::Unknown)?;
+                    return Ok((
+                        PreAuthorizationStatus::Success(
+                            transaction_intent_hash,
+                        ),
+                        delays,
+                    ));
+                }
+                SubintentStatus::Unknown => {
```

Comment:
I prefer as little code as possible inside of matches, especially when the number of lines is assymetrical - as in the variant `CommittedSuccess` returns quite early, but `Unknown` is 4x as many lines. Futhermore, we are in a tripple indendation here, a little bit pyramid of doom. But we note that we can just break out of the match in this variant, and continue the code outside of the match (but sill inside the loop), right? 

```rust
loop {
    match status {
        SubintentStatus::CommittesSuccess => { 
            ... return;
         }
        SubintentStatus::Unknown => { /* not done yet, continue outside of match */ }
    }
    // handle unknown here
    let accumulated_delay = delays.iter().sum::<u64>();
}